rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java

```diff
@@ -95,4 +95,14 @@ public static Optional<TransactionOutput> searchForOutput(List<TransactionOutput
             .filter(output -> output.getScriptPubKey().equals(outputScriptPubKey))
             .findFirst();
     }
+
+    public static Sha256Hash generateSigHashForP2SHInput(BtcTransaction btcTx, int inputIndex) {
+        TransactionInput input = btcTx.getInput(inputIndex);
+        Optional<Script> redeemScript = extractRedeemScriptFromInput(input);
+        if (redeemScript.isEmpty()) {
+            throw new IllegalArgumentException("Couldn't extract redeem script from p2sh input");
+        }
+
+        return btcTx.hashForSignature(inputIndex, redeemScript.get(), BtcTransaction.SigHash.ALL, false);
+    }
```

Comment:
This can be made more functional, assuming that btcTx cannot be null:

```java
public static Sha256Hash generateSigHashForP2SHInput(BtcTransaction btcTx, int inputIndex) {
    return Optional.ofNullable(btcTx.getInput(inputIndex))
        .flatMap(input -> extractRedeemScriptFromInput(input))
        .map(redeemScript -> btcTx.hashForSignature(inputIndex, redeemScript, BtcTransaction.SigHash.ALL, false)) 
        .orElseThrow(() -> new IllegalArgumentException("Couldn't extract redeem script from p2sh input"));
}
```

Reply:
updated! Thanks